haskell-api.md

```diff
@@ -0,0 +1,236 @@
+
+Therefore, I find the length concerns for `IN`-queries to be less problematic
+than our multi-id path-pieces of today.
+
+---
+
+## Response fields
+
+- Always include `id`s
+- Within reason, choose attributes directly for the work that is motivating the
+  endpoint
+- Include related resources in abbreviated form when Frontend needs extra data
+  about them. For example:
+  - `students: [{id:, firstName:, lastName:}]`
+  - `school: {id:, name:}`
+
+## Status Codes
+
+- 201 for creation
+- 204 for no content (i.e. our `Empty` responses)
+- 400s MUST follow our `ValidationError` machinery
+
+## Pagination
+
+- All list-returning routes of modest size should be paginated via `Yesod.Page`
+
+## Modules Structure
+
+- Fully expand all parents to start
+
+  This means we won't break Frontend when the tree expands over time
+
+- Fully name-space all parents and routes (see example)
+
+  **Caveats**:
+
+  - There's no need to `V3`-prefix everything
+  - Things can "start over" when there's the "single below a list" situation,
+    e.g. `TeacherP x $ TeacherR` instead of `TeachersP $ TeachersTeacherP x $
+    TeachersTeacherR`, which is a bit buffalo-buffalo
+  - The "start over" caveat does not apply to path naming (i.e. the above
+    example will still have the module `Teachers/Teacher.hs`)
+
+- Name modules to directly match parents and routes, which implies one Handler
+  module per route (see example)
+
+  **Caveats**:
+
+  - In reality, we need to prevent collisions between versions through
+    additional prefixing of route constructors. We ignore this in this guide.
+
+Example:
+
+```
+/3 V3P:                         --> Handlers/V3
+  /students StudentsP:          --> Handlers/V3/Students
+    / StudentsR GET             --> Handlers/V3/Students.hs
+    /#StudentId StudentP:       --> Handlers/V3/Students/Student
+      / StudentR GET            --> Handlers/V3/Students/Student.hs
+      /courses StudentCoursesP: --> Handlers/V3/Students/Student/Courses
+        / StudentCoursesR GET   --> Handlers/V3/Students/Student/Courses.hs
+```
+
+## Module Structure
+
+- Prefer in-module request/response types
+
+  - Request type naming: `{Resource}{Method}Body`
+  - Response type naming: `{Resource}{Method}`
+  - NOTE: list responses are `[{Resource}{Method}]`, not `{Resources}{Method}`
+
+  ```hs
+  -- Good
+  data TeacherGet
+
+  getTeachersR :: Handler [TeacherGet]
+  getTeachersR = undefined
+
+  data TeacherPost
+  data TeacherPostBody
+
+  postTeacherR :: Handler TeacherPost
+  postTeacherR = do
+    body <- requireJsonBody @TeacherPostBody
+
+  -- Bad
+  data Teachers
+
+  getTeachersR :: Handler Teachers
+  getTeachersR = undefined
+
+  data PostTeacher
+
+  postTeacherR :: Handler (Entity Teacher)
+  postTeacherR = do
+    body <- requireJsonBody @PostTeacher
+  ```
+
+- Prefer in-module data-access functions, until they require sharing
+- Order your module as:
+  - Request/response type(s)
```

Comment:
Does this include instances? We typically place class instances below the data type they are declaring for.

Reply:
> Does this include instances? 

This document makes no claim about that one way or the other, but...

> We typically place class instances below the data type they are declaring for

I agree with this, and it should probably go in our general guide. I'd consider that a normal Haskell thing that would be followed anywhere, including `Handler`s.

Reply:
Can a recommend flipping `handlers` and `Request/response types`? That order seems like the full progressive disclosure path.

Reply:
Hmm, I think that's too radical to do at this stage. Types first is a **strong** convention, followed generally in Haskell and very much in our own codebase. I'm a soft-no for that reason, but wouldn't block someone loudly in favor.

More importantly though, I wouldn't feel good about making that flip to get your Approve and then merging without the rest of the team re-weighing in on that specific point. Would you be willing to let this merge as-is and raising that ordering suggestion separately after?

Reply:
Yes, I'm fine to not be too controversial.

Reply:
I'm a huge proponent of progressive disclosure in module structure. I think for handler modules specifically, types first is still counts as a good form of progressive disclosure because generally the type name has the handler name in it (i.e. `TeachersGET`) and you see exactly what data will be returned up front. Then the handler function gives you details about how that is accomplished.

That's my 2 cents :man_shrugging:

Reply:
I can sympathize with that reason. The problem is that practically you get:

```hs
data TheType = _

instance ...

instance ...

instance ...

instance ...

theHandler = _
```

Reply:
That's a good point, I didn't consider the instances.